cardinal/search/composedsearch.go

```diff
@@ -0,0 +1,269 @@
+	var err error
+	for _, search := range orSearch.searches {
+		err = errors.Join(err, search.Each(eCtx, callback))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (orSearch *OrSearch) collect(eCtx engine.Context) ([]types.EntityID, error) {
+	resMap := make(map[types.EntityID]bool)
+	for _, search := range orSearch.searches {
+		ids, err := search.collect(eCtx)
+		if err != nil {
+			return nil, err
+		}
+		for _, id := range ids {
+			resMap[id] = true
+		}
+	}
+	res := make([]types.EntityID, 0)
+	for id := range resMap {
+		res = append(res, id)
+	}
+
+	return res, nil
+}
+
+func (orSearch *OrSearch) First(eCtx engine.Context) (types.EntityID, error) {
+	ids, err := orSearch.collect(eCtx)
+	if err != nil {
+		return 0, err
+	}
+	if len(ids) == 0 {
+		return 0, eris.New("No search results")
+	}
+	return ids[0], nil
+}
+
+func (orSearch *OrSearch) MustFirst(eCtx engine.Context) types.EntityID {
+	id, err := orSearch.First(eCtx)
+	if err != nil {
+		panic("no search results")
+	}
+	return id
+}
+
+func (orSearch *OrSearch) Count(eCtx engine.Context) (int, error) {
+	ids, err := orSearch.collect(eCtx)
+	if err != nil {
+		return 0, err
+	}
+	return len(ids), nil
+}
+
+func (andSearch *AndSearch) Each(eCtx engine.Context, callback CallbackFn) error {
+	ids := make(map[types.EntityID]int)
+	for _, search := range andSearch.searches {
+		subIDs, err := search.collect(eCtx)
+		if err != nil {
+			return err
+		}
+		for _, subid := range subIDs {
+			v, ok := ids[subid]
+			if !ok {
+				ids[subid] = 1
+			} else {
+				ids[subid] = v + 1
+			}
+		}
+	}
+	for k, v := range ids {
+		if v == len(andSearch.searches) {
+			if !callback(k) {
+				return nil
+			}
+		}
+	}
+	return nil
+}
+
+func (andSearch *AndSearch) collect(eCtx engine.Context) ([]types.EntityID, error) {
+	results := make([]types.EntityID, 0)
+	err := andSearch.Each(eCtx, func(id types.EntityID) bool {
+		results = append(results, id)
+		return false
+	})
+	if err != nil {
+		return nil, err
+	}
+	return results, nil
+}
+
+func (andSearch *AndSearch) First(eCtx engine.Context) (types.EntityID, error) {
+	ids, err := andSearch.collect(eCtx)
+	if err != nil {
+		return 0, err
+	}
+	if len(ids) == 0 {
+		return 0, eris.New("No search results")
+	}
+	return 0, nil
+}
+
+func (andSearch *AndSearch) MustFirst(eCtx engine.Context) types.EntityID {
+	id, err := andSearch.First(eCtx)
+	if err != nil {
+		panic("No search results")
+	}
+	return id
+}
+
+func (andSearch *AndSearch) Count(eCtx engine.Context) (int, error) {
+	ids, err := andSearch.collect(eCtx)
+	if err != nil {
+		return 0, err
+	}
+	return len(ids), nil
+}
+
+func (andSearch *AndSearch) evaluateSearch(eCtx engine.Context) []types.ArchetypeID {
+	searchCounts := make(map[types.ArchetypeID]int)
+	for _, search := range andSearch.searches {
+		ids := search.evaluateSearch(eCtx)
+		for _, id := range ids {
+			v, ok := searchCounts[id]
+			if !ok {
+				searchCounts[id] = 0
```

Comment:
Same as above, this can be simplified to `searchCounts[id]++`. It also fixes this off-by-one error here.